### PR TITLE
OSX fixes/definitions & OSX_BUILD=1

### PR DESCRIPTION
**Pull request by Hyenadae**
Added OSX build based off of my latest Master (Border fixes, removed N64 text, gfxpool fix etc).
Requires   'gmake' to compile
Closes https://github.com/sm64pc/sm64pc/issues/83

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -456,18 +472,25 @@ LDFLAGS := -lm -lGL -lSDL2 -no-pie -s TOTAL_MEMORY=20MB -g4 --source-map-base ht
...
 else
 
 # Linux / Other builds below
 ifeq ($(TARGET_RPI),1)
-LDFLAGS := $(OPT_FLAGS) -lm -lGLESv2 `$(CROSS)sdl2-config --libs` -no-pie
+LDFLAGS := $(OPT_FLAGS) -lm -lGLESv2 `$(SDLCONFIG) --libs` -no-pie
 else
-LDFLAGS := $(BITS) -march=$(TARGET_ARCH) -lm -lGL `$(CROSS)sdl2-config --libs` -no-pie -lpthread
+
+ifeq ($(TARGET_MACOS),1)
```
Comment (kemenaran) [inline Makefile:486]:
```suggestion
ifeq ($(OSX_BUILD),1)
```

**Comment by kemenaran** [comment]
`tools/Makefile:5` lacks an `-DOSX_BUILD` definition.

**Comment by Hyenadae** [comment]
Changed the name of that target_osx (patch had it...) to match everything else
version_cflags contains -DOSX_BUILD so the rest should be fine

**Comment by Xenthio** [comment]
merge it, this works

**Comment by Hyenadae** [comment]
Alright. Crap.

**Comment by Zoltar007** [comment]
I'm compiling this on MacOS 10.14.6.

`make OSX_BUILD=1` immediately fails with:
```
Makefile.split:152: build/us_pc/level_rules.mk: No such file or directory
cpp-9 -P -DVERSION_US -DNON_MATCHING -DAVOID_UB -I . -o build/us_pc/level_rules.mk levels/level_rules.mk
cc1: fatal error: opening output file build/us_pc/level_rules.mk: No such file or directory
compilation terminated.
make: *** [build/us_pc/level_rules.mk] Error 1
```
`gmake OSX_BUILD=1` doesn't encounter the above error, but fails at:
```
n64graphics_ci_dir/exoquant/exoquant.c:28:10: fatal error: 'malloc.h' file not
      found
#include <malloc.h>
```
Modifying "tools/n64graphics_ci_dir/exoquant/exoquant.c" as described in https://github.com/sm64pc/sm64pc/pull/92#commitcomment-39214716 does get past that error.

There are a whole lot of errors caused by lines 10 & 11 in `include/PR/os_libc.h`:
```
In file included from src/pc/ultra_reimplementation.c:3:
In file included from ./lib/src/libultra_internal.h:3:
In file included from include/ultra64.h:27:
include/PR/os_libc.h:10:13: error: expected parameter declarator
extern void bcopy(const void *, void *, size_t);
            ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_strings.h:46:52: note: 
      expanded from macro 'bcopy'
                __builtin___memmove_chk (dest, src, __VA_ARGS__, __darwi...
                                                                 ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_common.h:38:63: note: 
      expanded from macro '__darwin_obsz0'
#define __darwin_obsz0(object) __builtin_object_size (object, 0)
                                                              ^
In file included from src/pc/ultra_reimplementation.c:3:
In file included from ./lib/src/libultra_internal.h:3:
In file included from include/ultra64.h:27:
include/PR/os_libc.h:10:13: error: expected ')'
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_strings.h:46:52: note: 
      expanded from macro 'bcopy'
                __builtin___memmove_chk (dest, src, __VA_ARGS__, __darwi...
                                                                 ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_common.h:38:63: note: 
      expanded from macro '__darwin_obsz0'
#define __darwin_obsz0(object) __builtin_object_size (object, 0)
                                                              ^
include/PR/os_libc.h:10:13: note: to match this '('
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_strings.h:46:52: note: 
      expanded from macro 'bcopy'
                __builtin___memmove_chk (dest, src, __VA_ARGS__, __darwi...
                                                                 ^
In file included from src/pc/ultra_reimplementation.c:3:
In file included from ./lib/src/libultra_internal.h:3:
In file included from include/ultra64.h:27:
include/PR/os_libc.h:10:13: error: conflicting types for
      '__builtin___memmove_chk'
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_strings.h:46:3: note: 
      expanded from macro 'bcopy'
                __builtin___memmove_chk (dest, src, __VA_ARGS__, __darwi...
                ^
include/PR/os_libc.h:10:13: note: '__builtin___memmove_chk' is a builtin with
      type 'void *(void *, const void *, unsigned long, unsigned long)'
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_strings.h:46:3: note: 
      expanded from macro 'bcopy'
                __builtin___memmove_chk (dest, src, __VA_ARGS__, __darwi...
                ^
In file included from src/pc/ultra_reimplementation.c:3:
In file included from ./lib/src/libultra_internal.h:3:
In file included from include/ultra64.h:27:
include/PR/os_libc.h:11:13: error: expected parameter declarator
extern void bzero(void *, size_t);
            ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_strings.h:53:33: note: 
      expanded from macro 'bzero'
                __builtin___memset_chk (dest, 0, __VA_ARGS__, __darwin_o...
                                              ^
In file included from src/pc/ultra_reimplementation.c:3:
In file included from ./lib/src/libultra_internal.h:3:
In file included from include/ultra64.h:27:
include/PR/os_libc.h:11:13: error: expected ')'
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_strings.h:53:33: note: 
      expanded from macro 'bzero'
                __builtin___memset_chk (dest, 0, __VA_ARGS__, __darwin_o...
                                              ^
include/PR/os_libc.h:11:13: note: to match this '('
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_strings.h:53:26: note: 
      expanded from macro 'bzero'
                __builtin___memset_chk (dest, 0, __VA_ARGS__, __darwin_o...
                                       ^
In file included from src/pc/ultra_reimplementation.c:3:
In file included from ./lib/src/libultra_internal.h:3:
In file included from include/ultra64.h:27:
include/PR/os_libc.h:11:13: error: conflicting types for
      '__builtin___memset_chk'
extern void bzero(void *, size_t);
            ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_strings.h:53:3: note: 
      expanded from macro 'bzero'
                __builtin___memset_chk (dest, 0, __VA_ARGS__, __darwin_o...
                ^
include/PR/os_libc.h:11:13: note: '__builtin___memset_chk' is a builtin with
      type 'void *(void *, int, unsigned long, unsigned long)'
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_strings.h:53:3: note: 
      expanded from macro 'bzero'
                __builtin___memset_chk (dest, 0, __VA_ARGS__, __darwin_o...
                ^
```
Commenting out lines 10 & 11 in `include/PR/os_libc.h` does get past the above errors.

I am still seeing the following error:
```
In file included from src/pc/gfx/gfx_opengl.c:29:
/usr/local/include/SDL2/SDL_opengles2.h:35:10: fatal error: 
      'GLES2/gl2platform.h' file not found
#include <GLES2/gl2platform.h>
```
It seems that lines 26 to 30 in `src/pc/gfx/gfx_opengl.c` are the issue. Commenting them out except for line 27 does get past the error.

I am also still seeing the following error:
```
In file included from src/pc/gfx/gfx_sdl2.c:21:
/usr/local/include/SDL2/SDL_opengles2.h:35:10: fatal error: 
      'GLES2/gl2platform.h' file not found
#include <GLES2/gl2platform.h>
```
It seems that lines 18 to 22 are the issue. Commenting out all of them but line 19 gets past the error.

Making all of the above changes does produce a working binary.

**Comment by coltongit** [comment]
@Hyenadae mind building those fixes https://github.com/sm64pc/sm64pc/pull/92#issuecomment-629647944 has into this PR?

**Comment by Hyenadae** [comment]
Literally all the errors seem to be coming from the if statements not respecting OSX_BUILD being passed, even though it's defined in  VERSION_CFLAGS += -DOSX_BUILD
which gets passed around to all? files that get compiled

**Comment by Hyenadae** [comment]
Hey, I fixed a typo for strings.h, the Bzero / bcopy errors should be fixed.
But I'm still confused about malloc thing, will give the patch another look

**Comment by coltongit** [comment]
@Zoltar007 a fix just got pushed for a bunch of these errors you were having. Make sure you update, recompile, and tell us if your issues are fixed now

**Comment by Zoltar007** [comment]
As of commit 71c7890650a98f50674bb090aaea69ef8fe1f21f, I'm still seeing errors:
```
In file included from src/pc/ultra_reimplementation.c:3:
In file included from ./lib/src/libultra_internal.h:3:
In file included from include/ultra64.h:27:
include/PR/os_libc.h:10:13: error: expected parameter declarator
extern void bcopy(const void *, void *, size_t);
            ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_strings.h:46:52: note: 
      expanded from macro 'bcopy'
                __builtin___memmove_chk (dest, src, __VA_ARGS__, __darwi...
                                                                 ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_common.h:38:63: note: 
      expanded from macro '__darwin_obsz0'
#define __darwin_obsz0(object) __builtin_object_size (object, 0)
                                                              ^
In file included from src/pc/ultra_reimplementation.c:3:
In file included from ./lib/src/libultra_internal.h:3:
In file included from include/ultra64.h:27:
include/PR/os_libc.h:10:13: error: expected ')'
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_strings.h:46:52: note: 
      expanded from macro 'bcopy'
                __builtin___memmove_chk (dest, src, __VA_ARGS__, __darwi...
                                                                 ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_common.h:38:63: note: 
      expanded from macro '__darwin_obsz0'
#define __darwin_obsz0(object) __builtin_object_size (object, 0)
                                                              ^
include/PR/os_libc.h:10:13: note: to match this '('
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_strings.h:46:52: note: 
      expanded from macro 'bcopy'
                __builtin___memmove_chk (dest, src, __VA_ARGS__, __darwi...
                                                                 ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_common.h:38:54: note: 
      expanded from macro '__darwin_obsz0'
#define __darwin_obsz0(object) __builtin_object_size (object, 0)
                                                     ^
In file included from src/pc/ultra_reimplementation.c:3:
In file included from ./lib/src/libultra_internal.h:3:
In file included from include/ultra64.h:27:
include/PR/os_libc.h:10:13: warning: type specifier missing, defaults to 'int'
      [-Wimplicit-int]
extern void bcopy(const void *, void *, size_t);
            ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_strings.h:46:52: note: 
      expanded from macro 'bcopy'
                __builtin___memmove_chk (dest, src, __VA_ARGS__, __darwi...
                                                                 ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_common.h:38:32: note: 
      expanded from macro '__darwin_obsz0'
#define __darwin_obsz0(object) __builtin_object_size (object, 0)
                               ^
In file included from src/pc/ultra_reimplementation.c:3:
In file included from ./lib/src/libultra_internal.h:3:
In file included from include/ultra64.h:27:
include/PR/os_libc.h:10:13: error: conflicting types for
      '__builtin___memmove_chk'
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_strings.h:46:3: note: 
      expanded from macro 'bcopy'
                __builtin___memmove_chk (dest, src, __VA_ARGS__, __darwi...
                ^
include/PR/os_libc.h:10:13: note: '__builtin___memmove_chk' is a builtin with
      type 'void *(void *, const void *, unsigned long, unsigned long)'
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_strings.h:46:3: note: 
      expanded from macro 'bcopy'
                __builtin___memmove_chk (dest, src, __VA_ARGS__, __darwi...
                ^
In file included from src/pc/ultra_reimplementation.c:3:
In file included from ./lib/src/libultra_internal.h:3:
In file included from include/ultra64.h:27:
include/PR/os_libc.h:11:13: error: expected parameter declarator
extern void bzero(void *, size_t);
            ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_strings.h:53:33: note: 
      expanded from macro 'bzero'
                __builtin___memset_chk (dest, 0, __VA_ARGS__, __darwin_o...
                                              ^
In file included from src/pc/ultra_reimplementation.c:3:
In file included from ./lib/src/libultra_internal.h:3:
In file included from include/ultra64.h:27:
include/PR/os_libc.h:11:13: error: expected ')'
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_strings.h:53:33: note: 
      expanded from macro 'bzero'
                __builtin___memset_chk (dest, 0, __VA_ARGS__, __darwin_o...
                                              ^
include/PR/os_libc.h:11:13: note: to match this '('
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_strings.h:53:26: note: 
      expanded from macro 'bzero'
                __builtin___memset_chk (dest, 0, __VA_ARGS__, __darwin_o...
                                       ^
In file included from src/pc/ultra_reimplementation.c:3:
In file included from ./lib/src/libultra_internal.h:3:
In file included from include/ultra64.h:27:
include/PR/os_libc.h:11:13: error: conflicting types for
      '__builtin___memset_chk'
extern void bzero(void *, size_t);
            ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_strings.h:53:3: note: 
      expanded from macro 'bzero'
                __builtin___memset_chk (dest, 0, __VA_ARGS__, __darwin_o...
                ^
include/PR/os_libc.h:11:13: note: '__builtin___memset_chk' is a builtin with
      type 'void *(void *, int, unsigned long, unsigned long)'
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/secure/_strings.h:53:3: note: 
      expanded from macro 'bzero'
                __builtin___memset_chk (dest, 0, __VA_ARGS__, __darwin_o...
                ^
1 warning and 6 errors generated.
gmake: *** [Makefile:757: build/us_pc/src/pc/ultra_reimplementation.o] Error 1
```

**Comment by Zoltar007** [comment]
Changing line 7 in `include/PR/os_libc.h` from `#include <strings.h>` to `#include <_strings.h>` fixes the bzero & bcopy errors.

**Comment by Zoltar007** [comment]
Changing line 7 in `include/PR/os_libc.h` from `#include <strings.h>` to `#include <_strings.h>` doesn't fix the bzero & bcopy errors.

**Comment by Hyenadae** [comment]
> Changing line 7 in `include/PR/os_libc.h` from `#include <strings.h>` to `#include <_strings.h>` doesn't fix the bzero & bcopy errors. Even including both `<strings.h>` & `<_strings.h>`. The errors remain the same.

Strings doesn't fix the bcopy error, only the 'malloc' fix should'

**Comment by Zoltar007** [comment]
> > Changing line 7 in `include/PR/os_libc.h` from `#include <strings.h>` to `#include <_strings.h>` doesn't fix the bzero & bcopy errors. Even including both `<strings.h>` & `<_strings.h>`. The errors remain the same.
> 
> Strings doesn't fix the bcopy error, only the 'malloc' fix should'

Ah, ok. It seems that commit 71c7890650a98f50674bb090aaea69ef8fe1f21f didn't fix the issue.

**Comment by Hyenadae** [comment]
I think the issue is that the makefile doesn't pass these flags to the build section where tools and include is... as it's early on and separate from the main game (which happily accepts our compile flags)

**Comment by Zoltar007** [comment]
Looks like the issue is that `malloc.h` doesn't exist for MacOS. The equivalent is `stdlib.h`, which is already an include, so lines 27 to 29 can be removed from `tools/n64graphics_ci_dir/exoquant/exoquant.c`.

**Comment by Hyenadae** [comment]
I found part of the issue :/
There's *another* makefile within ./tools/ and it has it's own overrides for make. 
So we gotta make sure OSX_BUILD and it's defines get passed along so the code in the tools and include sections modify as expected.

**Comment by Hyenadae** [comment]
I have to go for a few hours. If anyone can make a way for the tools makefile to accept
make OSX_BUILD=1
(it'll change the compile output on make after a clean) 
then the existing code in my repo will work a lot better, if not properly now.

**Comment by coltongit** [comment]
@Zoltar007 can you please update and recompile? Another change might have fixed all your issues

**Comment by Zoltar007** [comment]
> @Zoltar007 can you please update and recompile? Another change might have fixed all your issues

It still fails with the same errors, unfortunately.

**Comment by Hyenadae** [comment]
@Zoltar007 Reclone and try again? Is anything different?

**Comment by Zoltar007** [comment]
> @Zoltar007 Reclone and try again? Is anything different?

It compiles without error as of commit 13375fd86a10a589cc60b5ed3d0b3f6725174150.
Commit a209ce8d8f7d2b30b1e1fd3a295b5ccc79965b78 allows it to be compiled with just `gmake` now.

**Comment by coltongit** [comment]
Glad to hear that your issues are solved, @Zoltar007. Enjoy!

**Comment by Zoltar007** [comment]
It may be best to revert the `shell uname` change in commit a209ce8d8f7d2b30b1e1fd3a295b5ccc79965b78 as it may break something in the future.

**Comment by Hyenadae** [comment]
Cool! I'm glad my last minute change before I had to run fixed it.
I'm gonna manually resync to a fresh repo and change all the files, should be cleaner for a pull.

**Pull request closed without merging**